src/ray/raylet/scheduling_policy.cc

```diff
@@ -165,37 +183,69 @@ bool SchedulingPolicy::ScheduleBundle(
   return resource_demand.IsSubset(available_node_resources);
 }
 
-std::vector<TaskID> SchedulingPolicy::SpillOver(
-    SchedulingResources &remote_scheduling_resources) const {
+std::vector<TaskID> SchedulingPolicy::SpillOverInfeasibleTasks(
+    SchedulingResources &node_resources) const {
   // The policy decision to be returned.
   std::vector<TaskID> decision;
-
-  ResourceSet new_load(remote_scheduling_resources.GetLoadResources());
+  ResourceSet new_load(node_resources.GetLoadResources());
 
   // Check if we can accommodate infeasible tasks.
   for (const auto &task : scheduling_queue_.GetTasks(TaskState::INFEASIBLE)) {
     const auto &spec = task.GetTaskSpecification();
     const auto &placement_resources = spec.GetRequiredPlacementResources();
-    if (placement_resources.IsSubset(remote_scheduling_resources.GetTotalResources())) {
+    if (placement_resources.IsSubset(node_resources.GetTotalResources())) {
       decision.push_back(spec.TaskId());
       new_load.AddResources(spec.GetRequiredResources());
     }
   }
+  node_resources.SetLoadResources(std::move(new_load));
+  return decision;
+}
 
+std::vector<TaskID> SchedulingPolicy::SpillOver(
+    SchedulingResources &remote_resources, SchedulingResources &local_resources) const {
+  // First try to spill infeasible tasks.
+  auto decision = SpillOverInfeasibleTasks(remote_resources);
+
+  // Get local available resources.
+  ResourceSet available_local_resources =
+      ResourceSet(local_resources.GetAvailableResources());
+  available_local_resources.SubtractResources(local_resources.GetLoadResources());
   // Try to accommodate up to a single ready task.
-  for (const auto &task : scheduling_queue_.GetTasks(TaskState::READY)) {
-    const auto &spec = task.GetTaskSpecification();
-    if (!spec.IsActorTask()) {
-      // Make sure the node has enough available resources to prevent forwarding cycles.
-      if (spec.GetRequiredPlacementResources().IsSubset(
-              remote_scheduling_resources.GetAvailableResources())) {
-        decision.push_back(spec.TaskId());
-        new_load.AddResources(spec.GetRequiredResources());
-        break;
+  bool task_spilled = false;
+  for (const auto &queue : scheduling_queue_.GetReadyTasksByClass()) {
+    // Skip tasks for which there are resources available locally.
+    const auto &task_resources =
+        TaskSpecification::GetSchedulingClassDescriptor(queue.first);
+    if (task_resources.IsSubset(available_local_resources)) {
+      continue;
+    }
```

Comment:
Won't this result in never spilling any tasks for where there are available local resources?

Reply:
I tried to mimic the same behavior as the other change in this PR. I think what you're saying is right, but it does take into account the current load from the ready queue. This change should result in spilling a task when the total load from the ready queue is greater than the available local resources.